Add NVMe-oF HII feature to timberland_1.0_final

trevor-cockrell opened the pull request:
3 commits introduced, divided by module:

- NetworkPkg
  - Addition of PcdMaxNvmeOfAttemptNumber
- NetworkPkg/Include
  - Addition of gNvmeOfConfigGuid
- NetworkPkg/NvmeOfDxe
  - Add HII/configuration forms
  - Add HII details in NvmeOfDxe README

trevor-cockrell committed:
NetworkPkg: Add Pcd for maximum NVMe-oF attempts

Add PcdMaxNvmeOfAttemptNumber to NetworkPkg as the maximum number of
NVMe-over-fabrics attempts.

Signed-off-by: Trevor cockrell <trevor_cockrell@dell.com>

trevor-cockrell committed:
NetworkPkg/Include: Add GUID gNvmeOfConfigGuid

GUID gNvmeOfConfigGuid added for NvmeOfDxe HII configuration forms.

Signed-off-by: Trevor cockrell <trevor_cockrell@dell.com>

trevor-cockrell committed:
NetworkPkg/NvmeOfDxe: NVMe-over-Fabrics configuration forms

Add support for modification of NVMe-oF connection attempts in the
NVMe-oF configuration submenu of the HII.
Update NvmeOfDxe readme with configuration form details.

Signed-off-by: Trevor cockrell <trevor_cockrell@dell.com>

johnmeneghini commented:
Where are the artifacts for this MR?  I'm looking for the downloadable .zip file.

trevor-cockrell commented:
> Where are the artifacts for this MR? I'm looking for the downloadable .zip file.

Had to dig -
try [artifact](https://github.com/timberland-sig/edk2/suites/13895685152/artifacts/772814781), from the [actions](https://github.com/timberland-sig/edk2/actions/runs/5390429943)

johnmeneghini commented on NetworkPkg/NvmeOfDxe/media/image048.png:
Can we pre-populate the `Host NQN` according to TP-TP-4126 ?

And there's not such thing at a `Host NID`. It's a `Host ID`

Douglas-Farley replied:
> From TP-4126 NVMe-oF Boot HostNQN and HostID: this is requirement for the host.

It's a `should`, not a `shall`.  I'm not opposed to following it, but it's not a requirement in that sense.

trevor-cockrell replied:
> And there's not such thing at a Host NID. It's a Host ID

Resolved in 5648f2e3826971f6b4898cfe4b71755b13ec05a1 and README updated

> Can we pre-populate the Host NQN according to TP-TP-4126 ?

Resolved in 0eb935b282b2eb6545f3ca3b72ca5b1ee46ba7af. Host ID and/or Host NQN will be pre-populated based on system UUID.

johnmeneghini reviewed (commented):
From TP-4126 NVMe-oF Boot HostNQN and HostID:  this is requirement for the host.
```
The pre-OS boot environment and OS environment should use a fixed platform UUID to create a HostNQN
 and HostID. The implementation should use the System UUID found in the SMBIOS table. The System
 Management BIOS (SMBIOS) Reference Specification is described in DSP0134. The SMBIOS table is
 typically available to pre-OS firmware and Expansion ROM Firmware in the pre-OS boot environment as
 well as to the OS environment. 
```

Douglas-Farley commented:
> From TP-4126 NVMe-oF Boot HostNQN and HostID: this is requirement for the host.

It's a `should`, not a `shall`.  I'm not opposed to following it, but it's not a requirement in that sense.

johnmeneghini commented:
> > From TP-4126 NVMe-oF Boot HostNQN and HostID: this is requirement for the host.
> 
> It's a `should`, not a `shall`. I'm not opposed to following it, but it's not a requirement in that sense.

Sorry Doug... I won't approve this change until it implements TP-4126.  It shouldn't be hard to do.

johnmeneghini commented:
Bug number 2:

The current code requires the gateway address to be programmed.  This is not currently required and I always leave the gateway blank in my attempt files.  There's nothing that says you have to have a gateway programmed on your subnet, especially with a static IP.

trevor-cockrell committed:
NetworkPkg/NvmeOfDxe: Convert Host NID to Host ID

Signed-off-by: Trevor cockrell <trevor_cockrell@dell.com>

trevor-cockrell committed:
NetworkPkg/NvmeOfDxe: Make host gateway optional

Signed-off-by: Trevor cockrell <trevor_cockrell@dell.com>

trevor-cockrell committed:
NetworkPkg/NvmeOfDxe: Populate Host NQN/ID according to TP-4126

During global data initialization: if empty, the Host NQN and/or
Host ID will be populated based on the system UUID.

Signed-off-by: Trevor cockrell <trevor_cockrell@dell.com>

trevor-cockrell commented:
> Bug number 2:
> 
> The current code requires the gateway address to be programmed. This is not currently required and I always leave the gateway blank in my attempt files. There's nothing that says you have to have a gateway programmed on your subnet, especially with a static IP.

Resolved in 2386e67d00e3418a6642c9469f1d0cbce90954bb 

Edit: New artifact [here](https://github.com/timberland-sig/edk2/suites/14777659657/artifacts/840561140)

trevor-cockrell committed:
NetworkPkg/NvmeOfDxe: Various fixes

  - Resolve MSVCC compilation errors with VS2019
  - NvmeOfConfig: fix shorter ipv6 addresses not converting from NV data
    to IFR data

Signed-off-by: Trevor cockrell <trevor_cockrell@dell.com>

WalonLi commented on NetworkPkg/NvmeOfDxe/NvmeOfMisc.c:
```diff
@@ -1824,3 +1914,127 @@ NvmeOfBeforeEBS (
+        DEBUG_ERROR,
+        "%a: Cannot read system GUID.\n",
+        __FUNCTION__,
+        &gNvmeOfConfigGuid
```
Only one parameter in this debug call, is that coding error?

trevor-cockrell replied:
Thanks! Resolved in [fa5d8f5](https://github.com/timberland-sig/edk2/pull/20/commits/fa5d8f541602b120fb5fd9c91a31097589874740)

mwilck commented:
This is great! I particularly like the fact that it's still possible to pre-populate the attempt data from the EFI disk image, sparing me from having to type those UUIDs :-)

Question: I tried to enable IPv6, but it says that that's unsupported. Is that intentional?
![image](https://github.com/timberland-sig/edk2/assets/15961354/60639c07-d8b5-4693-842a-a20a29414b8a)

If I hit ENTER at this point, it will switch back to IPv4, but the target IP address (which was an IPv4 address) will be cleared and must be re-entered.

mwilck commented:
I'm wondering about the "Autoconfigure" mode. If I select that, all other configuration settings including the target subsystem ID and target IP address go away. How is this mode supposed to work?

trevor-cockrell commented:
@mwilck, appreciate the feedback!!

> Question: I tried to enable IPv6, but it says that that's unsupported. Is that intentional?

No, however I've found where this is likely occurring. Fix incoming!

> If I hit ENTER at this point, it will switch back to IPv4, but the target IP address (which was an IPv4 address) will be cleared and must be re-entered.

Agreed - this is neither user-friendly nor desirable behavior. 😄 
Located issue in callback logic when switching between IP modes. Looking at reworking this to be a bit more robust/behave more intuitively.

> I'm wondering about the "Autoconfigure" mode. How is this mode supposed to work?

When `NvmeofIpMode` is set to `IP_MODE_AUTOCONFIG`, DHCP is enabled by the driver & attempted for both host and subsystem on IPv4 then IPv6 stacks. That being the case, I might have been a bit overzealous with suppressing fields in this mode.
Would it make more sense to show the (read-only) DHCP toggles in this mode as:
``` 
                                                       ^                        
   NVM Subsystem              <Disabled>                 IP4, IP6,              
                                                         Autoconfigure. In      
 > Network Device List                                   Autoconfigure mode,    
   NVMe-oF Network Device:    22:22:22:22:22:22          NVMe-oF driver will    
                                                         attempt to connect     
   Internet Protocol          <Autoconfigure>            NVMe-oF subsystem via  
                                                         IPv4 stack, if failed  
   Connection Retry Count     [3]                        then attempt IPv6      
                                                         stack.                 
   Enable DHCP                [X] // <- grayout for autoconfig
   Subsystem info from DHCP   [X] // <- grayout for autoconfig
                                                                                
   NVM Subsystem NID          7ac41e78-d9b0-4641-83c7-3d                        
                              f039ec24c6                                        
   Save Changes                                                                 
                                                       v                        
```

mwilck commented:
Thanks for explaining the autoconfig mode. I recall having seen it in the old config file as well, without understanding what it was supposed to mean.

Yes, showing the greyed-out selections for DHCP will help clarifying what this is about.

Side note: The name of the feature is poorly chosen IMO, because I'll probably not be the only one who would think of IPv6 address autoconfiguration (aka Router solicitation / advertisements) in the first place, with which our "autoconfiguration" has nothing to do.
Indeed I was wondering whether IPv6 autoconfiguration without DHCPv6 was supported, which doesn't seem to be the case.

Perhaps we should use the label "both IPv4 and IPv6" instead of "autoconnect"?

trevor-cockrell committed:
NetworkPkg/NvmeOfDxe: HII IPv6 check adjustment

Signed-off-by: Trevor Cockrell <trevor_cockrell@dell.com>

trevor-cockrell committed:
NetworkPkg/NvmeOfDxe: HII IP mode form tweaks

Grayout instead of suppress DHCP toggles in Autoconfigure mode

Signed-off-by: Trevor Cockrell <trevor_cockrell@dell.com>

johnmeneghini reviewed (dismissed):
Red Hat is satisfied with these changes.  Thanks for working on this.

johnmeneghini commented:
Internal testing of this branch has uncovered the following problems:

1. With empty OVMF_VARS.fd I found out that only the first configuration gets saved properly. When entering EFI Setup and trying to save changes for the second time, random values get lost. I was e.g. unable to save changes to the DHCP or SubsysNQN options in my boot attempt. Is this a known issue? 

2. I was having trouble saving  some other non-NBFT settings like the framebuffer resolution. Anyway,  made it working once configured fully on a first try.

3. Endianness conversion problem of the hostnqn. Consult this with the attached screenshot. 

```
# nvme nbft show --output-format=json
...
    "host":{
      "nqn":"nqn.2014-08.org.nvmexpress:uuid:12345678-abcd-abcd-abcd-
123456789012",
      "id":"78563412-cdab-cdab-abcd-123456789012",
    },
...

# cat /sys/class/nvme-fabrics/ctl/nvme0/hostnqn
nqn.2014-08.org.nvmexpress:uuid:12345678-abcd-abcd-abcd-123456789012
# cat /sys/class/nvme-fabrics/ctl/nvme0/hostid
78563412-cdab-cdab-abcd-123456789012
```

mwilck commented:
> Internal testing has found more bugs which need to be addressed.

I'd still recommend to merge this now, and address any bugs or other issues separately. The addition of the HII is a huge improvement in particular for new users, even if it isn't perfect yet.

mwilck reviewed (approved):
I approve this PR, even if the HII code doesn't work perfectly yet.

Douglas-Farley commented:
> Internal testing of this branch has uncovered the following problems:
>  [...]


John, just created a real bug for this so we can track it outside this PR in case we merge Hii and then run it to ground based on today's meeting.
see: https://github.com/timberland-sig/edk2/issues/23

johnmeneghini reviewed (approved):
As agreed at yesterdays meeting - Red Hat approves this PR.

trevor-cockrell commented:
Closing PR in favor of: https://github.com/timberland-sig/edk2/pull/24.
